### ISSUE TITLE
Possible to give Record types a name property?

### ISSUE THREAD

**Issue by Wideshanks** [plaid/sanctuary-def#7]
So instead of printing a list of all the fields as in:

https://github.com/plaid/sanctuary-def/blob/master/index.js#L293

--instead print the name of the Record in TypeError messages?

**Comment by davidchambers** [comment]
This is certainly worth discussing. I'd like to understand your motivations, as I have some questions:

  - Are you concerned about the verbosity of error messages for complex record types?

  - Should it even be valid to define two equivalent record types in an environment? For example:

    ```javascript
    const Coord = $.RecordType({x: $.Number, y: $.Number});
    const Point = $.RecordType({x: $.Number, y: $.Number});

    const def = $.create($.env.concat([Coord, Point]));
    ```

My thinking at the time was that `{ age :: Number, name :: String }` is more descriptive than `Person`. If an error message just referenced `Person`, one might need to consult code or documentation in order to learn the shape of a Person.

**Comment by Wideshanks** [comment]
> concerned about the verbosity of error messages for complex record types

Yeah, verbosity in (very) large records:

![scrolling-output](https://cloud.githubusercontent.com/assets/2061925/12061094/94c13966-af4b-11e5-9ac9-df1287b2cf8b.png)

Or maybe I should avoid using records when the data structures are that big?

**Comment by davidchambers** [comment]
> Or maybe I should avoid using records when the data structures are that big?

This seems to me like perfectly reasonable use of a record type. :)

It's interesting to note that in this case the representation of the record type accounts for only about 5% of the error message: the vast majority is devoted to the representation of the invalid return value.

I wonder whether the real issue is that the error messages do not explain *why* a value is invalid. Consider this example:

```javascript
const Point = $.RecordType({x: $.Number, y: $.Number});

const def = $.create($.env.concat([Point]));

//    dist :: Point -> Point -> Number
const dist =
def('dist',
    {},
    [Point, Point, $.Number],
    (a, b) => Math.sqrt(Math.pow(a.x - b.x, 2) + Math.pow(a.y + b.y, 2)));

dist({x: 0, y: 0}, {x: 3});
// ! TypeError: ‘dist’ expected a value of type { x :: Number, y :: Number } ↩︎
//              as its second argument; received {"x": 3}

dist({x: 0, y: 0}, {x: 3, y: '4'});
// ! TypeError: ‘dist’ expected a value of type { x :: Number, y :: Number } ↩︎
//              as its second argument; received {"x": 3, "y": "4"}
```

The record type defines only two fields, and there's no nesting, so it's easy to see why the second argument is invalid in each case. With more complex records, it would be useful for error messages to draw attention to erroneous fields. For example:

```javascript
dist({x: 0, y: 0}, {x: 3});
// ! TypeError: ‘dist’ expected a value of type { x :: Number, y :: Number } ↩︎
//              as its second argument; received {"x": 3} which lacks a ‘y’ field

dist({x: 0, y: 0}, {x: 3, y: '4'});
// ! TypeError: ‘dist’ expected a value of type { x :: Number, y :: Number } ↩︎
//              as its second argument; received {"x": 3, "y": "4"} whose ‘y’ field ↩︎
//              is of type String rather than Number
```

What do you think, @Wideshanks?

**Comment by Wideshanks** [comment]
Yeah, that looks perfect.

**Comment by davidchambers** [comment]
I just merged #55 which makes it clear *why* a given value is not a member of the specified record type.
